crates/handlers/src/graphql/mutations/user.rs

```diff
@@ -766,4 +860,154 @@ impl UserMutations {
             status: SetPasswordStatus::Allowed,
         })
     }
+
+    /// Register a user. If enabled, can be used by anonymous requesters to
+    /// create an account. May require a CAPTCHA challenge to be completed.
+    #[allow(clippy::too_many_lines)]
+    async fn register_user(
+        &self,
+        ctx: &Context<'_>,
+        input: RegisterUserInput,
+    ) -> Result<RegisterUserPayload, async_graphql::Error> {
+        let state = ctx.state();
+        let site_config = state.site_config();
+        if !site_config.password_registration_enabled {
+            return RegisterUserStatus::SelfRegistrationDisabled.into();
+        }
+
+        let activity_tracker = ctx.activity_tracker();
+        let http_client_factory = state.http_client_factory();
+        let url_builder = state.url_builder();
+        let mut repo = state.repository().await?;
+        let mut policy = state.policy().await?;
+        let clock = state.clock();
+        let homeserver = state.homeserver_connection();
+        let password_manager = state.password_manager();
+        let mut rng = state.rng();
+
+        // Validate the captcha
+        let passed_captcha = input
+            .captcha
+            .verify(
+                activity_tracker,
+                http_client_factory,
+                url_builder.public_hostname(),
+                site_config.captcha.as_ref(),
+            )
+            .await
+            .is_ok();
+
+        if !passed_captcha {
+            return RegisterUserStatus::InvalidCaptcha.into();
+        }
+
+        if input.username.is_empty() {
+            // TODO are there no other grammar restrictions???
```

Comment:
hmm, there is the `username_valid` function to validate locally? but then because we call `is_localpart_available`, the HS will validate that we're giving a good username